Modifications to VideoWrapper Video Storage

decodyng opened the pull request:
The goal of this PR is to clean up the way that videos are stored when a VideoWrapper is employed in the score_agent experiment. Specific things it does: 
- When `videos=True` in config, score_agent saves videos as Sacred artifacts
- When `video_dir=None` in config, score_agent saves videos in a tmpdir to avoid race conditions
- When `videos=True` in config and a FileStorageObserver is attached to the score_agent experiment, score_agent runs a cleanup operation on saved videos, pulling them into a `videos` file and sorting by `env` number subfolders 
- Modifies VideoWrapper to fail when you pass it a directory that already exists 
- Adds video-specific test within test_experiments, and confirmed they, and all other tests in test_experiments, pass. 

Note for @AdamGleave, per our conversation: I currently have the tmpdir being created on the score_agent level, and being passed into VideoWrapper (rather than have the logic happen within VideoWrapper itself), because this allows you to actually have access to the tmpdir location for the purposes of saving sacred artifacts. This is a little annoying, because it makes the change to VideoWrapper more of a breaking change, and also makes the tmpdir logic less general, but I wasn't sure how else to make the tmpdir information visible on the experiment level. Definitely open to hearing whether you think the current setup of this makes sense. I could also do something like adding a parameter to VideoWrapper such that if, say, `tmpdir=True`, a tmpdir is autocreated within VideoWrapper, meaning that in cases where the calling code doesn't care about the files being stored in a particular non-Sacred place, that code doesn't need to handle the tmpdir logic.

AdamGleave commented on src/modelfree/common/utils.py:
```diff
@@ -2,6 +2,7 @@
 import datetime
 import itertools
 import os
+import shutil
```
unused import

AdamGleave replied:
I suggest running `ci/code_checks.sh` before commit which will catch this (and some other linting errors). I have it installed as a Git commit hook.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -2,7 +2,12 @@
 
 import functools
 import os.path as osp
-
```
The linter is going to yell at you for this not being sorted alphabetically

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -49,6 +54,37 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+    if len(video_files) == 0:
+        return
+
+    ptn = re.compile('env_(\d*)_video.(\d*)')
```
I think this should be using Python's raw string syntax: r''. Otherwise the `\d` could be interpreted as escape sequences.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -105,9 +149,28 @@ def env_fn(i):
+            for video_file_path in os.listdir(env_video_dir):
+                if 'mp4' in video_file_path:
+                    env_number = env_video_dir.split("/")[-1]
+                    print("Env number: {}".format(env_number))
```
Remove or turn into a logger.debug statement

decodyng replied:
Ah, yep, that should have just gotten removed

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -105,9 +149,28 @@ def env_fn(i):
+                    score_ex.add_artifact(filename=os.path.join(env_video_dir, video_file_path),
+                                          name=sacred_name)
+        except:
+            Warning("Can't find path {}; no videos from that path added as artifacts".format(env_video_dir))
```
I don't think this does anything? You probably want `warnings.warn`.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -63,10 +99,11 @@ def default_score_config():
-    episodes = 20                       # number of episodes to evaluate
-    render = True                       # display on screen (warning: slow)
+    episodes = 1                        # number of episodes to evaluate
+    render = False                      # display on screen (warning: slow)
```
I view the common case as being interactive usage of score_agent so would vote for keeping `render=True` as default. It's easy enough to set `render = False` as being default in batch scripts etc.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -63,10 +99,11 @@ def default_score_config():
     num_env = 1                         # number of environments to run in parallel
-    episodes = 20                       # number of episodes to evaluate
-    render = True                       # display on screen (warning: slow)
+    episodes = 1                        # number of episodes to evaluate
```
Why just one episode?

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -119,5 +182,6 @@ def main():
     score_ex.run_commandline()
 
 
+
```
too many lines of whitespace

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -49,6 +54,37 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+def _clean_video_directory_structure(observer_obj):
+    """
+    A simple utility method to take saved videos within a Sacred run structure and clean up the file pathways, so that
+    all videos are organized under a "videos" directory, and within that organized by environment ID.
```
I think it's actually fine to have videos with different environment IDs in the same directory (but would require renumbering them). They should all be from the same distribution: we just run environments in parallel as an optimization.

decodyng replied:
Sounds good; I wasn't quite sure whether that was true, but I'm happy for this to be simpler

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -119,5 +182,6 @@ def main():
 
+
 if __name__ == '__main__':
     main()
```
Consider adding to `main()` a logging statement that outputs the Sacred directory for the `FileStorageObserver`? I'm imagining a common use case being running it and then wondering which of the ~100 Sacred directories you should be looking at...

decodyng replied:
I like this idea; added as a log statement

decodyng commented:
Changed video directory to not sort out videos in per-environment subdirectories, added more logging statements, fixed linting issues. 

@AdamGleave Could you take another look?

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -13,6 +19,8 @@
 
 score_ex = Experiment('score')
 
+logging.basicConfig(format='%(levelname)s - %(message)s', level=logging.DEBUG)
```
Put this in `main()`.

If I import `score_agent` from another file, I don't want it to override my logging config.

AdamGleave replied:
Also Sacred already supports some logging config, e.g. you can specify the logging level with the `-l` flag. I'm OK with us overriding this if there's a compelling reason though, but would like to try and centralize logging in one place.

decodyng replied:
I didn't realize Sacred had it's own config (more accurately, I'm still figuring out the nuances of how logging configs waterfall); it works perfectly fine to just remove this line

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +58,33 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+    and within that organized by environment ID.
+
+    :param observer_obj: A Sacred FileStorageObserver object
+    :return:
```
Returns None?

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +58,33 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+    """
+    A simple utility method to take saved videos within a Sacred run structure and clean
+    up the file pathways, so that all videos are organized under a "videos" directory,
+    and within that organized by environment ID.
```
Is it organized by environment ID now?

decodyng replied:
Nope, fixed

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +58,33 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+            continue
+        episode_id = search_result.groups()[0]
+        new_file_name = "episode_{}_recording.mp4".format(int(episode_id))
+        os.rename(video_file, os.path.join(new_video_dir, new_file_name))
```
Episode IDs won't be unique so I think this is missing some logic. Maybe number it as `env_id * num_envs + episode_id`, or just number them sequentially...?

Also, should we copy over the `*.meta.json` file as well? Sometimes we stick useful fields in there, e.g. `winners`.

decodyng replied:
Ah, good point. I'm a little uncomfortable about inferring an ordered/sequential episode numbering in the multi env case where there isn't actually a natural/guaranteed ordering. I think I'd just prefer to add the env ID to the file name

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -66,7 +101,11 @@ def default_score_config():
     videos = False                      # generate videos
-    video_dir = 'data/videos/'          # video directory
+    video_dir = None                    # directory to store videos in.
+
```
delete newline

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -75,6 +114,15 @@ def default_score_config():
                 record_traj, record_traj_params, num_env, episodes, render, videos, video_dir):
+    if videos:
+        if video_dir is None:
+            logging.info("""No directory provided for saving videos; using a tmpdir instead,
```
For just two-lines I'd probably use `"first line" "second line"` instead of triple quotes, but no strong preference.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -105,18 +154,40 @@ def env_fn(i):
     observer = FileStorageObserver.create(osp.join('data', 'sacred', 'score'))
     score_ex.observers.append(observer)
     score_ex.run_commandline()
+    logging.info("Sacred run completed, files stored at {}".format(observer.dir))
```
I tend to favour having `logger = logging.getLogger('module.name')` at the top level, and then call `logger` rather than `logging`. This lets you have per-module logging levels (e.g. debug in one module, info everywhere else) and will include the module name in logging output. Doesn't matter too much when it's a CLI script, but really useful for library code.

decodyng replied:
I initially decided not to do that because it didn't seem like there was a clean way to have the same logger object visible in both the main() method and in the sacred experiment, unless we defined it globally, which I've done now.

decodyng commented:
@AdamGleave Made changes to logging; reorganized code to save metadata files and clean up main score_agent function.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -12,6 +18,7 @@
 from modelfree.envs.gym_compete import GymCompeteToOurs, game_outcome
 
 score_ex = Experiment('score')
+score_ex_logger = logging.getLogger(__name__)
```
Tempting but `__name__` is equal to `__main__` when running interactively which probably isn't what you want. I've just been hardcoding it.

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +57,62 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+    new_video_dir = os.path.join(basedir, "videos")
+    os.mkdir(new_video_dir)
+    new_video_metadata_dir = os.path.join(new_video_dir, "metadata")
+    os.mkdir(new_video_metadata_dir)
```
You could use `os.makedirs(new_video_metadata_dir)` to do this and the previous `os.mkdir` in one fell swoop, but I'm not sure that's actually better.

decodyng replied:
Yeah, I think I like being a bit more procedural about it, even if it technically could be condensed

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +57,62 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+    if video_search_result is not None:
+        episode_id = video_search_result.groups()[0]
+        sacred_name = "env_{}_episode_{}_recording.mp4".format(env_number, int(episode_id))
+
```
delete whitespace

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +57,62 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+    elif metadata_search_result is not None:
+        episode_id = metadata_search_result.groups()[0]
+        sacred_name = "env_{}_episode_{}_metadata.json".format(env_number, int(episode_id))
+
```
delete whitespace

decodyng replied:
I'll make the change for the sake of getting the branch merged rather than going back and forth on it, but I actually think that having whitespace in cases like this is better/makes the code more readable than the alternative

AdamGleave commented on src/modelfree/score_agent.py:
```diff
@@ -50,6 +57,62 @@ def get_empirical_score(_run, env, agents, episodes, render=False):
+
+    else:
+        return
+    score_ex.add_artifact(filename=os.path.join(env_dir, saved_video_path),
```
add newline
